### DISCUSSION TITLE
How to connect the SCXML data model with External Data Storage

### DISCUSSION THREAD

**Discussion by tfahringer** [alexzhornyak/SCXML-tutorial#27] [Q&A]
I am interested to access external data repositories for instance Amazon S3 from SCXML.
The SCXML data model allows to declare variables which then can be used in operations of states, transitions, etc.

One can open Amazon S3 and generated, read and write objects. How could the SCXML data model extended for this opportunity to access data in external repositories. One could refer to data objects/variables in S3 and then use these variables in SCXML operations wherever normal SCXML variables can be used as well.

Is this foreseen by the syntax of SCXML?

**Comment by alexzhornyak** [comment]
[SCXML standard](https://www.w3.org/TR/scxml/#profiles) allows to use custom datamodels but its definition depends on SCXML framework (uscxml, Qt scxml, scion etc.). I mean you will do it on the framework side

The definition of a data model must:

- Specify the boolean expression language used as the value of the 'cond' attribute in \<transition\>, \<if\> and \<elseif\> This language must not have side effects and must include the predicate 'In', which takes a single argument, the id of a state in the enclosing state machine, and returns 'true' if the state machine is in that state.
- Specify the location expression language that is used as the value of the 'location' attribute of the \<assign\> tag.
- Specify the value expression language that is used as the value of the 'expr' attribute of the \<data\> and \<assign\> elements.
- Specify the scripting language used inside the \<script\> element

You can watch [here examples of different datamodels for uscxml framework](https://github.com/tklab-tud/uscxml/tree/master/src/uscxml/plugins/datamodel)
